infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/ERC20BurnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+import "../../Xai.sol";
+import "../../upgrades/referee/Referee9.sol";
+import "../../upgrades/node-license/NodeLicense8.sol";
+import "../../upgrades/pool-factory/PoolFactory3.sol";
+
+/**
+ * @title esXai
+ * @dev Implementation of the esXai
+ */
+contract esXai4 is ERC20Upgradeable, ERC20BurnableUpgradeable, AccessControlUpgradeable {
+
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    EnumerableSetUpgradeable.AddressSet private _whitelist;
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+    address public xai;
+    bool private _redemptionActive;
+    mapping(address => RedemptionRequest[]) private _redemptionRequests;
+    address public esXaiBurnFoundationRecipient;
+    uint256 public esXaiBurnFoundationBasePoints;
+    mapping(address => RedemptionRequestExt[]) private _extRedemptionRequests;
+    address public refereeAddress;
+    address public nodeLicenseAddress;
+    uint256 public maxKeysNonKyc;
+    address public poolFactoryAddress;
+
+    bool private _notEntered;
+    mapping(address => uint256) public _totalPendingRedemptions;
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[492] private __gap;
+
+    struct RedemptionRequest {
+        uint256 amount;
+        uint256 startTime;
+        uint256 duration;
+        bool completed;
+    }
+
+    struct RedemptionRequestExt {
+        uint256 amount;
+        uint256 startTime;
+        uint256 duration;
+        uint256 endTime;
+        bool completed;
+        bool cancelled;
+        bool voucherIssued;
+        uint256[5] __gap;
```

Comment:
Can you confirm that the gap does not have to be updated here ?

Reply:
Confirmed the contracts compile without changing the gap.

Reply:
Compile is not an issue when changing storage layount, they will always compile if there are no syntax errors.

The storage layout will only be verified rom hardhat when you want to deploy an upgrade, hardhat won't let you unless you explicitly ignore storage layout errors.

So you either have to try and deploy the upgrade to sepolia or run the upgrade tests in the smart contract test suite, we should have these test contracts just for sotrage validation, let me know if you know what I mean

Reply:
Understood, will do.